[WIP] v23.1.12 Release notes (#18020)
cockroachdb/docs: src/current/_includes/releases/v23.1/v23.1.12.md

## v23.1.12

Release Date: November 1, 2023

{% include releases/release-downloads-docker-image.md release=include.release %}

<h3 id="v23-1-12-security-updates">Security updates</h3>

- The `SIGHUP` signal now clears the cached expiration times for [client certificates](https://www.cockroachlabs.com/docs/v23.1/cockroach-cert#how-security-certificates-work) that are reported by the `security.certificate.expiration.client` metric. [#111006][#111006]
- SQL commands that were previously limited to the `admin` [system privilege](https://www.cockroachlabs.com/docs/v23.1/security-reference/authorization#supported-privileges) can now be used by users with the `VIEWCLUSTERMETADATA` or `REPAIRCLUSTERMETADATA` system privilege, depending on whether the operation is read-only or modifies state. [#111131][#111131]

<h3 id="v23-1-12-general-changes">General changes</h3>

- Fixed a bug where, internally, if we print a 0 decimal with a very low exponent we use excessive memory. This is not possible when using the [DECIMAL](https://www.cockroachlabs.com/docs/v23.1/decimal.md) type, but may be possible when using `crdb_internal` functions. [#110564][#110564]

> - Previously, when memory was exceeded while executing the [`EXPORT INTO PARQUET`](https://www.cockroachlabs.com/docs/v23.1/export#export-a-table-into-parquet) statement, this resulted in [out-of-memory crashes (OOMs)](https://www.cockroachlabs.com/docs/v23.1/cluster-setup-troubleshooting#out-of-memory-oom-crash). Now, if memory is exceeded, the `EXPORT INTO PARQUET` statement will return an error. If you see an error related to memory, retry the `EXPORT INTO PARQUET` statement using a smaller value for the [`chunk_rows` option](https://www.cockroachlabs.com/docs/v23.1/export#export-options). Cockroach Labs [recommends using changefeeds to export data](https://www.cockroachlabs.com/docs/v23.1/export-data-with-changefeeds) because they provide better performance for growing workloads. Additionally, changefeeds [operate as jobs](https://www.cockroachlabs.com/docs/v23.1/show-jobs), which offer [observability](https://www.cockroachlabs.com/docs/v23.1/monitor-and-debug-changefeeds) and [job management](https://www.cockroachlabs.com/docs/v23.1/create-and-configure-changefeeds). [#110717][#110717]

Author: Done

- New datetime built-ins (`make_date`, `make_timestamp`, and `make_timestamptz`) have been added, allowing for the creation of timestamps, timestamps with time zones, and dates. In addition, `date_trunc` now allows for a timestamp to be truncated in a specified timezone (to a specified precision). [#110338][#110338]
- New system privileges `CREATEROLE`, `CREATELOGIN`, `CREATEDB`, and `CONTROLJOB` have been introduced. Each is analogous to its existing [role option](https://www.cockroachlabs.com/docs/v23.1/create-user#role-options) counterpart but can additionally be inherited through role membership. [#110359][#110359], [#110359][#110359], [#110359][#110359], [#110359][#110359]
- [`RESTORE`](https://www.cockroachlabs.com/docs/v23.1/restore) can now be passed a `WITH EXECUTION LOCALITY` option similar to `BACKUP` to restrict execution of the job to nodes with matching localities. [#110611][#110611]
- The `statement_activity` and `transaction_activity` table column `execution_total_cluster_seconds` is now accurate. The `combinedstmts` endpoint returns the correct value for the `StmtsTotalRuntimeSecs` and `TxnsTotalRuntimeSecs` properties. [#109639][#109639]

> - SQL commands that were previously limited to the `admin` [system privilege](https://www.cockroachlabs.com/docs/v23.1/security-reference/authorization#supported-privileges) can now be used by users with the `VIEWCLUSTERMETADATA` or `REPAIRCLUSTERMETADATA` system privilege, depending on whether the operation is read-only or modifies state. [#111058][#111058]

Reviewer: I'm unsure if this should be SQL or Security. It seems like it's duplicated above.

> - The `discard` [log message](https://www.cockroachlabs.com/docs/v23.1/logging-overview) is now limited to once per minute by default. The message now includes both the number of transactions and the number of statements that were discarded. [#110983][#110983]

Reviewer: Should this be Observability?

Author: Hmm the script didn't make an Observability section. Let me make one.

Author: I don't see an Observability section in any of the previous release notes 🤔

Reviewer: You're right, it looks like they usually get put into DB Console changes.

- The new backup option `updates_cluster_monitoring_metrics` tracks the timestamp of the last backup failure due to a KMS error. This option is disabled by default. [#111310][#111310]

> - Fixed a bug where vectorized `COPY FROM` could produce a plan with more than one RenderNodes, when only zero or one should be allowed. This could result in multiple render nodes in a table with a hash sharded primary key. [#111412][#111412]

Reviewer: Bug fixes?

> - Fixed a bug that caused CockroachDB to stop collecting new statistics about [Statement fingerprints](https://www.cockroachlabs.com/docs/v23.1/ui-statements-page#statement-fingerprint-page) and [Transaction fingerprints](https://www.cockroachlabs.com/docs/v23.1/ui-transactions-page). [#111641][#111641]

Reviewer: Bug fixes?

> - Reduced memory usage in [SQL activity](https://www.cockroachlabs.com/docs/cockroachcloud/statements-page) jobs. [#112323][#112323]

Reviewer: Observability?

<h3 id="v23-1-12-operational-changes">Operational changes</h3>

> - - The [`cockroach debug zip`](https://www.cockroachlabs.com/docs/v23.1/cockroach-debug-zip.md) command now has an option to omit goroutine stack dumps. This impacts the creation of `nodes/*/stacks.txt` and `nodes/*/stacks_with_labels.txt` within debug ZIP bundles. Users can opt to exclude these goroutine stacks by using the `--include-goroutine-stacks=false` flag.  Note that fetching stack traces for all goroutines is a "stop-the-world" operation, which can momentarily have negative impacts on SQL service latency. Note also that any periodic goroutine dumps previously taken on the node will still be included in `nodes/*/goroutines/*.txt.gz`, as these would have already been generated and don't require any stop-the-world operations. [#110266][#110266]

Reviewer: ```suggestion
- The [`cockroach debug zip`](https://www.cockroachlabs.com/docs/v23.1/cockroach-debug-zip.md) command now has an option to omit goroutine stack dumps. This impacts the creation of `nodes/*/stacks.txt` and `nodes/*/stacks_with_labels.txt` within debug ZIP bundles. Users can opt to exclude these goroutine stacks by using the `--include-goroutine-stacks=false` flag.  Fetching stack traces for all goroutines is a "stop-the-world" operation, which can momentarily but significantly increase SQL service latency. Any periodic goroutine dumps previously taken on the node will still be included in `nodes/*/goroutines/*.txt.gz`, since they have already been generated and don't require any stop-the-world operations. [#110266][#110266]
```

- Requests for database details or table details from the UI, or usages of `SHOW RANGES WITH DETAILS` are no longer subject to errors if the number of requested spans is too large. [#109902][#109902]

> - Added a new [metric](https://www.cockroachlabs.com/docs/v22.2/metrics) `changefeed.lagging_ranges` that shows the number of [ranges](https://www.cockroachlabs.com/docs/v22.2/architecture/overview#architecture-range) which are behind in changefeeds. This metric can be used with the [`metrics_label` changefeed option](https://www.cockroachlabs.com/docs/v22.2/create-changefeed#options). The calculation of this metric is controlled by 2 new [cluster settings](https://www.cockroachlabs.com/docs/v22.2/cluster-settings): (1) `changefeed.lagging_ranges_threshold`, with a default of 3 minutes, is the amount of time a range needs to be behind to be considered lagging, and (2) `changefeed.lagging_ranges_polling_interval`, with a default of 1 minute, controls how often the lagging ranges calculation is done. Note that polling adds latency to the metric being updated. For example, if a range falls behind by 3 minutes, the metric may not update until an additional minute afterwards. Also note that ranges undergoing an initial scan for longer than the threshold are considered to be lagging. Starting a changefeed with an initial scan on a large table will likely increment the metric for each range in the table. However, as ranges complete the initial scan, the number of lagging ranges will decrease. [#110963][#110963]

Reviewer: Observability or Backup?

- The [DB Console](https://www.cockroachlabs.com/docs/v23.1/ui-overview) now constructs client-side requests using relative URLs instead of absolute ones. This enables proxying of the DB Console at arbitrary subpaths. [#111652][#111652]

> - Added a new [cluster setting](https://www.cockroachlabs.com/docs/v23.1/cluster-settings) `server.http.base_path` that controls the redirection of the browser after successful login with [OIDC SSO](https://www.cockroachlabs.com/docs/v23.1/sso-db-console). It is unlikely that this setting would need adjustment. However, it is helpful in cases where CockroachDB is running behind a load balancer or proxy that serves CockroachDB under a subpath, such as `https:// <hostname>/crdb/ `. In those cases, it is necessary for the browser to redirect to `/ crdb` after login instead of `/`, which has always been the hard-coded default.[#112038][#112038]

Reviewer: ```suggestion
- Added a new [cluster setting](https://www.cockroachlabs.com/docs/v23.1/cluster-settings) `server.http.base_path` that controls the redirection of the browser after successful login with [OIDC SSO](https://www.cockroachlabs.com/docs/v23.1/sso-db-console). Most users do not need to modify this setting. However, it is helpful in cases where CockroachDB is running behind a load balancer or proxy that serves CockroachDB under a subpath, such as `https:// <hostname>/crdb/ `. In those cases, it is necessary for the browser to redirect to `/ crdb` after login instead of `/`, which has always been the hard-coded default.[#112038][#112038]
```

<h3 id="v23-1-12-command-line-changes">Command-line changes</h3>

> - Added limited `statement_statistics` to the debug zip. [#110662][#110662]

Reviewer: I'd put this near the other debug.zip note and delete this section.

<h3 id="v23-1-12-db-console-changes">DB Console changes</h3>

- Non-admin users can now view the [**Database**, **Database Details**, and **Database table** pages](https://www.cockroachlabs.com/docs/v23.1/ui-databases-page). [#110342][#110342]
- The [SQL Connection Rate metric](https://www.cockroachlabs.com/docs/v23.1/ui-sql-dashboard#sql-connection-rate) on the [SQL Dashboard](https://www.cockroachlabs.com/docs/v23.1/ui-sql-dashboard) is downsampled using the MAX function instead of SUM. This improves situations where zooming out would cause the connection rate to increase for downsampled data. [#110497][#110497]

> - Fixed a bug in DB Console's [Statement Diagnostic page](https://www.cockroachlabs.com/docs/cockroachcloud/statements-page#diagnostics) that could cause the page to crash if the response was larger than 50 KB. The page now keeps pulling results until no maximum size errors are encountered. [#111278][#111278]

Reviewer: Bug fixes or leave it here? I'm mixed.

> - Fixed an error on the [SQL Activity page](https://www.cockroachlabs.com/docs/v23.1/ui-overview#sql-activity) when there was a workload, and then the workload stopped so that no queries ran against the database in the last hour. [#111496][#111496]

Reviewer: Bug fixes or leave it here? I'm mixed.

> - The [**Jobs** table](https://www.cockroachlabs.com/docs/v23.1/ui-jobs-page) will now correctly display timestamps for creation, last modified, and the completed time fields. [#111901][#111901]

Reviewer: Bug fixes or leave it here? I'm mixed.

<h3 id="v23-1-12-bug-fixes">Bug fixes</h3>

- Fixed a nil dereference panic during node startup that could be caused by an incorrect initialization order. [#109684][#109684]
- Fixed a bug where the [`BEGIN` statement](https://www.cockroachlabs.com/docs/v23.1/begin-transaction) incorrectly displayed a transaction ID in telemetry logs. It now correctly shows no transaction ID, since there is no open transaction when `BEGIN` is executed.[#109840][#109840]
- Fixed a bug that could cause a transaction performing multiple parallel foreign key checks to return a `concurrent txn use detected` error. [#109849][#109849]
- Fixed a bug where dependencies on sequences from tables would be reported with the wrong value for the `classid` column in the `pg_catalog.pg_depend` table. [#110207][#110207]

> - The `Reset Sql Stats` button is now visible to users with the admin role on the DB Console. [#110256][#110256]

Reviewer: DB Console section?

- Fixed edge cases in decimal and float evaluation for division operators. `'NaN'::DECIMAL / 0` will now return `NaN` instead of a division-by-zero error, and `0 / 'inf'::DECIMAL` will return `0` instead of `0E-2019`. [#110296][#110296]
- Fixed a DB Console issue where the `DROP_UNUSED` index recommendations produced by the table details page produced an invalid `DROP INDEX` statement. [#110453][#110453]
- Executing two [`ALTER RANGE default CONFIGURE ZONE`](https://www.cockroachlabs.com/docs/v23.1/alter-range#configure-zone) statements on the same line no longer triggers an error. [#110337][#110337]

> - Fixed a buggy TTL descriptor repair by removing it. Previously, upgrading from v22.2.x to 23.1.9 incorrectly removed [TTL storage parameters](https://www.cockroachlabs.com/docs/v23.1/row-level-ttl#ttl-storage-parameters) from tables (visible via [`SHOW CREATE TABLE  <ttl-table>`](https://www.cockroachlabs.com/docs/v23.1/show-create)) while attempting to repair table descriptors. This resulted in the node that attempts to run the TTL [job](https://www.cockroachlabs.com/docs/v23.1/show-jobs.html) crashing due to a panic caused by the missing TTL storage parameters. Clusters currently on v22.2.x should **not** be upgraded to v23.1.9 and should be upgraded directly to v23.1.10 or later. For more information, see [Technical Advisory 110363](../advisories/a110363.html). [#110500][#110500]

Reviewer: ```suggestion
- Removed a buggy TTL descriptor repair. Previously, upgrading from v22.2.x to 23.1.9 incorrectly removed [TTL storage parameters](https://www.cockroachlabs.com/docs/v23.1/row-level-ttl#ttl-storage-parameters) from tables (visible via [`SHOW CREATE TABLE  <ttl-table>`](https://www.cockroachlabs.com/docs/v23.1/show-create)) while attempting to repair table descriptors. A node that attempts to run the TTL [job](https://www.cockroachlabs.com/docs/v23.1/show-jobs.html) could crash due to a panic caused by the missing TTL storage parameters. Clusters currently on v22.2.x should **not** be upgraded to v23.1.9, but should move directly to v23.1.10 or later. For more information, refer to [Technical Advisory 110363](../advisories/a110363.html). [#110500][#110500]
```
I think the link to the TA should be the full https:// link.

> - Fixed a bug causing performance regression when disabling `sql.metrics.statement_details.enabled` which caused execution stats to be collected for all queries instead of the default one percent. [#109881][#109881]

Reviewer: ```suggestion
- Fixed a performance regression when disabling `sql.metrics.statement_details.enabled`, which caused execution stats to be collected for all queries instead of the default 1% of queries. [#109881][#109881]
```

> - `cockroach debug pebble` commands now work correctly with encrypted stores which don't use the default `cockroach-data` path without having to also pass `--store`. [#110507][#110507]

Reviewer: ```suggestion
- `cockroach debug pebble` commands now work correctly with encrypted stores which don't use the default `cockroach-data` path, and you no longer need to pass the `--store` option. [#110507][#110507]
```

> - The [**Database** pages](https://www.cockroachlabs.com/docs/cockroachcloud/databases-page) will no longer display `undefined` regions and outdated node information. [#110741][#110741]

Reviewer: ```suggestion
- The [**Database** pages](https://www.cockroachlabs.com/docs/cockroachcloud/databases-page) no longer displays `undefined` regions or outdated node information. [#110741][#110741]
```

- Fixed a bug where `RESET (ttl_expire_after)` could incorrectly remove `ttl_expiration_expression`. [#110746][#110746]

> - Fixed a bug where an [`ALTER TABLE ... ADD CONSTRAINT CHECK ...`](https://www.cockroachlabs.com/docs/v23.1/alter-table#add-constraint) statement that utilized a user-defined function in the `CHECK` could cause a validation error. [#110720][#110720]

Reviewer: ```suggestion
- Fixed a bug where an [`ALTER TABLE ... ADD CONSTRAINT CHECK ...`](https://www.cockroachlabs.com/docs/v23.1/alter-table#add-constraint) statement with a user-defined function (UDF) in the `CHECK` could cause a validation error. [#110720][#110720]
```

- Fixed a bug where `CREATE INDEX` for [partial indexes](https://www.cockroachlabs.com/docs/v23.1/partial-indexes) could fail with `ERROR: duplicate key value violates unique constraint` if concurrent inserts happened simultaneously. [#110584][#110584]
- Fixed a bug where a [`CREATE TABLE`](https://www.cockroachlabs.com/docs/v23.1/create-table) command with an `IDENTITY` column did not properly propagate the type of the column into the sequence. [#111014][#111014]
- Fixed a bug where the `format_type` built-in did not honor `typemod` information for array types, leading to incorrect output. [#110940][#110940]
- Fixed compaction behavior to prevent rapid sublevel growth when removing replicas from a store, such as during decommissioning. [#111141][#111141]

> - Fixed a bug that could occur when the "multiple active portals" execution mode (Preview) was enabled to evaluate queries such as lookup joins. The bug could result in an internal error like `unexpected 40960 leftover bytes` if the portal was not fully consumed. [#110666][#110666]

Reviewer: ```suggestion
- Fixed a bug that could occur when the "multiple active portals" execution mode (Preview) was enabled to evaluate queries such as lookup joins. The bug could result in an internal error like `unexpected 40960 leftover bytes` if the portal was not fully consumed. [#110666][#110666]
```
Link to somewhere for "multiple active portals"? Google didn't help me. Maybe there aren't docs.

- External connection URLs now accept the scheme `azure-blob` for connections to Azure Blob Storage and the scheme `azure-kms` for connections to Azure KMS. For backward compatibility, schemes `azure` and `azure-storage` schemes continue to work for connections to Azure Blob Storage. [#111246][#111246]
- Fixed a bug where changing the setting `server.telemetry.hot_ranges_stats.interval` had no effect. [#111373][#111373]

> - Added a check for values before using `mean` on the [Plan Details page](https://www.cockroachlabs.com/docs/v23.1/ui-statements-page), fixing a crash. [#111505][#111505]

Reviewer: ```suggestion
- Added a check for values before using `mean` on the DB Console [Plan Details page](https://www.cockroachlabs.com/docs/v23.1/ui-statements-page), fixing a crash. [#111505][#111505]
```
I wonder if it is worth grouping "like" bugs, so having all the DB Console ones near each other, for example.

- Fixed a bug where dependencies on sequences from tables would be reported with the wrong value for the `classid` column in the `pg_catalog.pg_depend` table. [#111600][#111600]

> - Fixed a bug that could cause a transaction performing multiple parallel foreign key checks to return a `concurrent txn use detected` error. [#111600][#111600]

Reviewer: This is duplicated above.

> - Removed buggy [TTL](https://www.cockroachlabs.com/docs/v23.1/row-level-ttl) descriptor repair. Previously, upgrading from v22.2.X to v23.1.9 incorrectly removed TTL storage parameters from tables (visible by running a `SHOW CREATE TABLE <ttl-table>;` statement) while attempting to repair table descriptors. This resulted in the node that attempted to run the TTL job crashing due to a panic caused by the missing TTL storage parameters. [#111600][#111600]

Reviewer: This is duplicated above (and I gave suggestions on it). Maybe choose one place.

- Fixed a bug where `atttypmod` in `pg_attribute` was not populated for [`TIMESTAMP`](https://www.cockroachlabs.com/docs/v23.1/timestamp) / [`INTERVAL`](https://www.cockroachlabs.com/docs/v23.1/interval) types, which meant that ORMs could not know the precision of these types properly. [#111727][#111727]

> - Fixed a bug in [transaction insight details](https://www.cockroachlabs.com/docs/v23.1/ui-insights-page) where it was possible to see the contention details of other transactions. Now, CockroachDB will only surface contention details for the current transaction. [#111880][#111880]

Reviewer: ```suggestion
- Fixed a bug in the DB Console [Transaction Insight Details](https://www.cockroachlabs.com/docs/v23.1/ui-insights-page) page , which showed  contention details of other transactions. Now, CockroachDB will only surface contention details for the current transaction. [#111880][#111880]
```

> - Fixed a bug where `indoption` inside `pg_index` was not properly encoded causing clients to be unable to decode it as `int2vector`. [#111957][#111957]

Reviewer: ```suggestion
- Fixed a bug where `indoption` inside `pg_index` was not properly encoded. Clients were unable to decode it as `int2vector`. [#111957][#111957]
```

> - CockroachDB now properly handles RPC failures on writes using the parallel commit protocol that execute in parallel to the commit operation, avoiding incorrect retryable failures and `transaction unexpectedly committed` assertions by detecting when writes cannot be retried idempotently, instead returning an `AmbiguousResultError`. [#111876][#111876]

Reviewer: ```suggestion
- RPC failures on writes now use the parallel commit protocol and execute in parallel to the commit operation. This change prevents incorrect retryable failures and `transaction unexpectedly committed` assertions by detecting when writes cannot be retried idempotently, instead returning an `AmbiguousResultError`. [#111876][#111876]
```

> - Fixed an issue where the [optimizer](https://www.cockroachlabs.com/docs/v23.1/cost-based-optimizer) fails to honor the `statement_timeout` session setting when generating constrained index scans for queries with large `IN` lists or `= ANY` predicates on multiple index key columns, which may lead to an out of memory condition on the node. [#112076][#112076]

Reviewer: ```suggestion
- Fixed a bug that prevented the [optimizer](https://www.cockroachlabs.com/docs/v23.1/cost-based-optimizer) from honoring the `statement_timeout` session setting when generating constrained index scans for queries with large `IN` lists or `= ANY` predicates on multiple index key columns. This bug could cause an Out-of-Memory (OOM) condition on the node. [#112076][#112076]
```

- Fixed a bug that caused internal errors during query optimization in rare cases. The bug has been present since version v2.1.11, but it is more likely to occur in version v21.2.0 and later, though it is still rare. The bug only presents when a query contains `min` and `max` [aggregate functions](https://www.cockroachlabs.com/docs/v23.1/functions-and-operators#aggregate-functions) [#112254][#112254]

> - Fixed an issue where lookup and index joins on tables with at least 3 column families were sometimes incorrectly evaluated, leading to "non-nullable column with no value" errors or incorrect query results. The bug was introduced in 22.2. [#113107][#113107]

Reviewer: ```suggestion
- Fixed a bug  where a lookup or index join on a table with at least three column families could be evaluated incorrectly, leading to a "non-nullable column with no value" error or incorrect query results. The bug was introduced in v22.2. [#113107][#113107]
```

<h3 id="v23-1-12-performance-improvements">Performance improvements</h3>

- Queries that compare collated strings now use less memory and may execute faster. [#110147][#110147]
- Reduced the impact of high-concurrency blind writes to the same key on goroutine scheduling latency. [#109370][#109370]

<div class="release-note-contributors" markdown="1">

<h3 id="v23-1-12-contributors">Contributors</h3>

This release includes 207 merged PRs by 64 authors.
We would like to thank the following contributor from the CockroachDB community:

> - craig (first-time contributor)

Reviewer: I don't think this is a real first-time contributor -- can you confirm?

Author: @mdlinville I looked at the AUTHORS file and didn't see their name. Is there any other place I should look at?

Reviewer: I think "craig" is a bot that we use. https://github.com/cockroachdb/cockroach/pulls/app%2Fcraig

Author: ohhh 😄

- Fixed a bug where executing the [`EXPORT INTO PARQUET`](https://www.cockroachlabs.com/docs/v23.1/export#export-a-table-into-parquet) statement could cause an [out-of-memory crash (OOM)](https://www.cockroachlabs.com/docs/v23.1/cluster-setup-troubleshooting#out-of-memory-oom-crash). Now, if memory is exceeded, the `EXPORT INTO PARQUET` statement returns an error. If you see an error related to memory, retry the `EXPORT INTO PARQUET` statement using a smaller value for the [`chunk_rows` option](https://www.cockroachlabs.com/docs/v23.1/export#export-options). Cockroach Labs [recommends using changefeeds to export data](https://www.cockroachlabs.com/docs/v23.1/export-data-with-changefeeds) because they provide better performance for growing workloads. Additionally, changefeeds [operate as jobs](https://www.cockroachlabs.com/docs/v23.1/show-jobs), which offer [observability](https://www.cockroachlabs.com/docs/v23.1/monitor-and-debug-changefeeds) and [job management](https://www.cockroachlabs.com/docs/v23.1/create-and-configure-changefeeds). [#110717][#110717] [#110717][#110717]

> - Added limited `statement_statistics` to the debug zip. [#110662][#110662]

Reviewer: ```suggestion
- Added limited `statement_statistics` to the `debug.zip` output. [#110662][#110662]
```
